base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2016 Daniel Reis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api, tools
+
+# Extended name search is only used on some operators
+ALLOWED_OPS = {'ilike', 'like'}
+
+
+@tools.ormcache(skiparg=0)
+def _get_rec_names(self):
+    "List of fields to search into"
+    model = self.env['ir.model'].search(
+        [('model', '=', self._name)])
+    rec_name = [self._rec_name] if bool(self._rec_name) else []
+    other_names = model.name_search_ids.mapped('name')
+    return rec_name + other_names
+
+
+def _extend_name_results(self, domain, results, limit):
```

Comment:
def _extend_name_results(self, domain, results, limit):
    result_count = len(results)
    if result_count < limit:
    if not limit:
        domain += [('id', 'not in', [x[0] for x in results])]
        recs = self.search(domain)
        results.extend(recs.name_get())
    elif result_count < limit:
        domain += [('id', 'not in', [x[0] for x in results])]
        recs = self.search(domain, limit=limit - result_count)
        results.extend(recs.name_get())